lib/internal/streams/end-of-stream.js

```diff
@@ -79,10 +79,24 @@ function eos(stream, opts, callback) {
   };
 
   const onclose = () => {
-    if (readable && !readableEnded) {
-      callback.call(stream, new ERR_STREAM_PREMATURE_CLOSE());
-    } else if (writable && !writableFinished) {
-      callback.call(stream, new ERR_STREAM_PREMATURE_CLOSE());
+    if (readable) {
+      const ended = (stream._readableState &&
+        stream._readableState.endEmitted) || stream.readableEnded;
+      if (!ended && !readableEnded) {
+        callback.call(stream, new ERR_STREAM_PREMATURE_CLOSE());
+      } else if (!readableEnded) {
+        // Compat. Stream has ended but haven't emitted 'end'.
+        callback.call(stream);
+      }
+    } else if (writable) {
+      const finished = (stream._writableState &&
+        stream._writableState.finished) || stream.writableFinished;
```

Comment:
ws.finished in not the same as ws.ended is it? the old compat was to ignore premature close if the stream had been ended (ie ws.end() was called)

Reply:
@mafintosh: correct, the workaround for that is that pipeline ignores that premature error in a safe way. See the changes in pipeline.

Reply:
That way we get "correct" behaviour from eos while not breaking and still having "correct" behaviour in pipeline.

Reply:
I see, okay, let me try to grok that a bit.

Thinking out loud it would be really nice to check if a stream is autoDestroy'ing (which I think at some point should be the default) and then simply wait for `close` always as the terminating event. Should be much simpler.

Reply:
Then we could keep the compat mode around for "old" streams but autoDestroying stream would run down the simple path.

Reply:
I'm not sure I follow this. Did you want something changed and if so can you describe it further?

Reply:
@mafintosh Any further groking?

Reply:
My point was that is a stream has autoDestroy set to true, which is probably gonna be the default behaviour at some some, we just need to check if close has been emitted so all of this becomes much easier.

I *do* think that's worth adding relatively soon.